src/RateLimiter/Counter.php

```diff
@@ -0,0 +1,140 @@
+     * @param int $period period to apply limit to, in seconds
+     * @param CacheInterface $storage
+     */
+    public function __construct(int $limit, int $period, CacheInterface $storage)
+    {
+        if ($limit < 1) {
+            throw new \InvalidArgumentException('The limit must be a positive value.');
+        }
+
+        if ($period < 1) {
+            throw new \InvalidArgumentException('The period must be a positive value.');
+        }
+
+        $this->limit = $limit;
+        $this->period = $period * self::MILLISECONDS_PER_SECOND;
+        $this->storage = $storage;
+
+        $this->incrementInterval = (float)($this->period / $this->limit);
+    }
+
+    public function setId(string $id): void
+    {
+        $this->id = $id;
+    }
+
+    public function getCacheKey(): string
+    {
+        return self::ID_PREFIX . $this->id;
+    }
+
+    public function incrementAndGetResult(): CounterStatistics
+    {
+        if ($this->id === null) {
+            throw new \LogicException('The counter ID should be set');
+        }
+
+        $this->lastIncrementTime = $this->getCurrentTime();
+        $theoreticalNextIncrementTime = $this->calculateTheoreticalNextIncrementTime(
+            $this->getLastStoredTheoreticalNextIncrementTime()
+        );
+        $remaining = $this->calculateRemaining($theoreticalNextIncrementTime);
+        $resetAfter = $this->calculateResetAfter($theoreticalNextIncrementTime);
+
+        if ($remaining >= 1) {
+            $this->storeTheoreticalNextIncrementTime($theoreticalNextIncrementTime);
+        }
+
+        return new CounterStatistics($this->limit, $remaining, $resetAfter);
+    }
+
+    /**
+     * @param float $storedTheoreticalNextIncrementTime
+     * @return float theoretical increment time that would be expected from equally spaced increments at exactly rate limit
+     * In GCRA it is known as TAT, theoretical arrival time.
+     */
+    private function calculateTheoreticalNextIncrementTime(float $storedTheoreticalNextIncrementTime): float
+    {
+        return max($this->lastIncrementTime, $storedTheoreticalNextIncrementTime) + $this->incrementInterval;
+    }
+
+    /**
+     * @param float $theoreticalNextIncrementTime
+     * @return int the number of remaining requests in the current time period
+     */
+    private function calculateRemaining(float $theoreticalNextIncrementTime): int
+    {
+        $incrementAllowedAt = $theoreticalNextIncrementTime - $this->period;
+
+        return (int)(round($this->lastIncrementTime - $incrementAllowedAt) / $this->incrementInterval);
+    }
+
+    private function getLastStoredTheoreticalNextIncrementTime(): float
+    {
+        return $this->storage->get($this->getCacheKey(), (float)$this->lastIncrementTime);
+    }
+
+    private function storeTheoreticalNextIncrementTime(float $theoreticalNextIncrementTime): void
+    {
+        $this->storage->set($this->getCacheKey(), $theoreticalNextIncrementTime);
```

Comment:
Set ttl to preserve leak.
`Ttl=$theoreticalNextincrementTime - time() +1`

Reply:
What do you mean by "to preserve leak"?

Reply:
Creates a cache record for each `$id` that never expires.

If the `$id` has many variant, the cache will easily be full of 5-10 thousand or more records that will never be deleted. However, in all likelihood, these records are never needed at the same time.

Reply:
Right. Thanks.

Reply:
Done.